src/hotspot/share/prims/universalUpcallHandler.cpp

```diff
@@ -37,18 +37,37 @@
 
 extern struct JavaVM_ main_vm;
 
-JavaThread* ProgrammableUpcallHandler::maybe_attach_and_get_thread(bool* should_detach) {
+// When an upcall is invoked from a thread that is not attached to the VM, we need to attach it,
+// and then to detach it at some point later. Detaching a thread as soon as the upcall completes
+// is suboptimal, as the same thread could later upcall to Java again, at which point the VM would
+// create multiple Java views of the same native thread. For this reason, we use thread local storage
+// to keep track of the fact that we have attached a native thread to the VM. When the thread local
+// storage is destroyed (which happens when the native threads is terminated), we check if the
+// storage has an attached thread and, if so, we detach it from the VM.
+struct UpcallContext {
+  Thread* attachedThread;
+
+  ~UpcallContext() {
+    if (attachedThread != NULL) {
+      JavaVM_ *vm = (JavaVM *)(&main_vm);
+      vm->functions->DetachCurrentThread(vm);
+    }
+  }
+};
+
+
+APPROVED_CPP_THREAD_LOCAL UpcallContext threadContext;
+
+JavaThread* ProgrammableUpcallHandler::maybe_attach_and_get_thread() {
   JavaThread* thread = JavaThread::current_or_null();
   if (thread == nullptr) {
     JavaVM_ *vm = (JavaVM *)(&main_vm);
     JNIEnv* p_env = nullptr; // unused
     jint result = vm->functions->AttachCurrentThread(vm, (void**) &p_env, nullptr);
     guarantee(result == JNI_OK, "Could not attach thread for upcall. JNI error code: %d", result);
```

Comment:
Do you really want to abort if something goes wrong i.e. out-of-memory**? (Also the JNI error code won't be very illuminating as only JNI_ERR is ever returned from AttachCurrentThread.)

** Any unexpected exceptions from the java.lang.Thread construction will be silently cleared but also cause failure.

Reply:
For errors that occurs inside the upcall, we generally try to print a stack trace - but in this case we're a bit of a limbo between native and Java - and we still don't have a Java thread - so I'm assuming that if something goes wrong here there's not much we can do to dress it up nicely and/or recover? I agree that printing out JNI error code seems over the top (and not very useful).

I note that this code has not changed, even in the foreign-preview branch, which contains a bunch of JVM cleanups. Maybe we can file a followup (but I'd like to understand better what you are proposing - and what options we have on the table).

Reply:
I tend to agree - only observation is that reporting the JNI error code, as David noted, is probably useless.

Reply:
I admit I don't know what the full context is here but at some level you have a piece of external code calling something and expecting something useful to happen as a result. I would have expected that API, whatever it is, to have some way of indicating that an error occurred and that the "something useful" did not in fact happen. If the initial code does not even know it is calling into a JVM I would expect it to be pretty miffed when the JVM decides to blow away the entire process due to an incidental issue the JVM had! 

Even if aborting is your only option I don't think a guarantee, with ensuing JVM core dump and hs_err file, is the way to go about it. I would not want to be fielding bug reports against the JVM in such a case.

Cheers.

Reply:
I think aborting is the only option. The flow starts with some native call made by Java code. At some point the native code upcalls back into Java, and, for some reason, attaching a thread fails. When this failure occurs we are deep inside a native library. Popping back until we reach some Java frame seems undesirable (what happens to all the native code in the middle?). As I pointed out yesterday, we have a similar situation when the *Java* code inside an upcall fails with some exception: but in that case at least we have a thread, so we can dump a stack trace, which makes it "look" like a normal exception (but it's not, in the sense that it cannot be caught by the Java code - as the VM exits after dumping the stack trace).
This is worse, as we can't even dump the stack trace (as we have no thread), so the best we can do is to dump some message, and then exit. I agree that generating a VM dump is too much, and we should scale that back. But I don't think there is a meaningful way to recover (as the recovery would have to happen inside the C code, which Panama cannot, by definition, control).